pkg/k8s/exposer.go

```diff
@@ -2,131 +2,114 @@
+	entityName := fmt.Sprintf("%s-%s", app.Peer, app.Name)
+	for _, managedResource := range exp.managedResources {
+		addErr := managedResource.Add(k8sResourceMetadata{
+			entityName:       entityName,
+			initialApp:       app,
+			childReturnedApp: addedApp,
+		}, clientset)
+		if addErr != nil {
+			return peers.App{}, multierr.Combine(addErr, exp.child.Withdraw(app))
+		}
 	}
+	return peers.WithAddress(addedApp, fmt.Sprintf("%s.%s:%d", entityName, exp.namespace, app.OriginalPort)), nil
+}
 
-	serviceName := fmt.Sprintf("%s-%s", app.Peer, app.Name)
-	service := &corev1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      serviceName,
-			Namespace: factory.namespace,
-			Labels:    factory.buildLabelsForSvc(),
-			Annotations: map[string]string{
-				"x-wormhole-app":  app.Name,
-				"x-wormhole-peer": app.Peer,
-			},
-		},
-		Spec: corev1.ServiceSpec{
-			Ports: []corev1.ServicePort{{
-				Port:       app.OriginalPort,
-				TargetPort: intstr.FromInt(port),
-			}},
-			Selector: factory.ownSelectors,
-		},
-	}
-	var upsertErr error
-	previousService, getErr := servicesClient.Get(context.Background(), serviceName, metav1.GetOptions{})
-	if errors.IsNotFound(getErr) {
-		logrus.Debugf("Creating service %s", serviceName)
-		_, upsertErr = servicesClient.Create(context.Background(), service, metav1.CreateOptions{})
-	} else if getErr != nil {
-		return peers.App{}, multierr.Combine(
-			fmt.Errorf("Could not get service %s: %v", serviceName, getErr), factory.Withdraw(addedApp),
-		)
-	} else {
-		logrus.Debugf("Updating service %s", serviceName)
-		service.SetResourceVersion(previousService.GetResourceVersion())
-		_, upsertErr = servicesClient.Update(context.Background(), service, metav1.UpdateOptions{})
+func (exp *k8sResourceExposer) Withdraw(app peers.App) error {
+	clientset, clientSetErr := exp.clientProvider.New()
+	if clientSetErr != nil {
+		return clientSetErr
 	}
-	if upsertErr != nil {
-		return peers.App{}, multierr.Combine(
-			fmt.Errorf("Unable to upsert the service: %v", upsertErr), factory.Withdraw(addedApp),
-		)
+	entityName := fmt.Sprintf("%s-%s", app.Peer, app.Name)
+	for i := range exp.managedResources {
+		managedResource := exp.managedResources[len(exp.managedResources)-1-i]
+		removeErr := managedResource.Remove(entityName, clientset)
+		if removeErr != nil {
+			return removeErr
+		}
 	}
-	return peers.WithAddress(addedApp, fmt.Sprintf("%s.%s:%d", serviceName, factory.namespace, app.OriginalPort)), nil
+	return exp.child.Withdraw(app)
 }
 
-func (factory *k8sServiceExposer) Withdraw(app peers.App) error {
-	serviceName := fmt.Sprintf("%s-%s", app.Peer, app.Name)
-	logrus.Debugf("Deleting service %s", serviceName)
-	config, inClusterConfigErr := rest.InClusterConfig()
-	if inClusterConfigErr != nil {
-		return inClusterConfigErr
-	}
-	clientset, clientSetErr := kubernetes.NewForConfig(config)
+func (exp *k8sResourceExposer) WithdrawAll() error {
```

Comment:
Is it used anywhere else then along tests? What's the usecase for such withdrawal? Part of graceful shutdown of whole wormhole on the specific cluster?

Reply:
Good point, it's unused. I wanted to use it when wormhole starts to cleanup all remnants of old state. https://github.com/glothriel/wormhole/issues/76